### PR TITLE
Driver hashes rsa v21

### PR DESCRIPTION
**Pull request by mpg**
Implement #6098 

**Gatekeeping notes:** no ChangeLog entry (part of a series, see #6146), no backport (new feature).

This PR also constitutes an example of how to adapt crypto modules (not governed by `MBEDTLS_USE_PSA_CRYPTO`) so that they can work in builds with driver-only hashes and no MD, by using PSA as a fall-back when MD is not present, to ensure backwards compatibility.

### CODE REVIEW

File: include/mbedtls/mbedtls_config.h [hunk 1]
```diff
@@ -1141,7 +1141,10 @@
  *
  * Enable support for PKCS#1 v2.1 encoding.
  *
- * Requires: MBEDTLS_MD_C, MBEDTLS_RSA_C
+ * Requires: MBEDTLS_RSA_C and (MBEDTLS_MD_C or MBEDTLS_PSA_CRYPTO_C).
```
Comment (AndrzejKurek) [inline include/mbedtls/mbedtls_config.h:1144]: Should this be
```suggestion
 * Requires: MBEDTLS_RSA_C and (MBEDTLS_MD_C or ( MBEDTLS_PSA_CRYPTO_C and PSA_WANT_ALG_MD5) ).
```
?

File: library/rsa.c [hunk 2]
```diff
@@ -54,6 +54,18 @@
 #include <stdlib.h>
 #endif
 
+/* We use MD first if it's available (for compatibility reasons)
+ * and "fall back" to PSA otherwise (which needs psa_crypto_init()). */
+#if defined(MBEDTLS_PKCS1_V21)
+#if defined(MBEDTLS_MD_C)
+#define HASH_MAX_SIZE   MBEDTLS_MD_MAX_SIZE
+#else /* MBEDTLS_MD_C */
+#include "psa/crypto.h"
```
Comment (AndrzejKurek) [inline library/rsa.c:63]: So we're OK with RSA using PSA even if `MBEDTLS_USE_PSA_CRYPTO` is not defined?

File: library/rsa.c [hunk 3]
```diff
@@ -502,10 +514,8 @@ int mbedtls_rsa_set_padding( mbedtls_rsa_context *ctx, int padding,
     if( ( padding == MBEDTLS_RSA_PKCS_V21 ) &&
         ( hash_id != MBEDTLS_MD_NONE ) )
     {
-        const mbedtls_md_info_t *md_info;
-
-        md_info = mbedtls_md_info_from_type( hash_id );
-        if( md_info == NULL )
+        /* Just make sure this hash is supported in this build. */
+        if( mbedtls_hash_info_get_size( hash_id ) == 0 )
```
Comment (AndrzejKurek) [inline library/rsa.c:518]: Minor:
```suggestion
        if( mbedtls_hash_info_psa_from_md( hash_id ) == PSA_ALG_NONE )
```
seems more appropriate

File: library/rsa.c [hunk 4]
```diff
@@ -1088,29 +1098,68 @@ int mbedtls_rsa_private( mbedtls_rsa_context *ctx,
 }
 
 #if defined(MBEDTLS_PKCS1_V21)
+#if !defined(MBEDTLS_MD_C)
+static int ret_from_status( psa_status_t status )
```
Comment (AndrzejKurek) [inline library/rsa.c:1102]: This function has future potential to be refactored to `psa_to_mbedtls_error( psa_status_t status, int module )` to have it uniform across modules.

File: library/rsa.c (hunk 2 above)
Comment (mprse) [reply library/rsa.c:63]: If `MBEDTLS_PKCS1_V21` is defined then one of `MBEDTLS_MD_C`, `MBEDTLS_PSA_CRYPTO_C` must be defined. Otherwise we have build error:

https://github.com/Mbed-TLS/mbedtls/pull/6141/files#diff-4fc3811c7c53e1746e1056ac38a6462ad2160077a59898bbd2bc31f5938ac185L165

Comment (AndrzejKurek) [reply library/rsa.c:63]: Yes, these can be defined, but `MBEDTLS_USE_PSA_CRYPTO` can be undefined to signal that we're not using PSA for (as its doc says) `x509 and TLS library, and enable new APIs for using keys handled by PSA Crypto`. I'm just wondering if it's readable that PSA can be used for something even though `MBEDTLS_USE_PSA_CRYPTO` is undefined.

**Review by mprse — COMMENTED** [review]
I did a quick review and looks good at first sight (no critical issues).

But it seems that rebase is now needed.

File: library/rsa.c (hunk 4 above)
Comment (mpg) [reply library/rsa.c:1102]: Good point! Could come in handy for #6146 and #6147. What would be a good home for this function? Would `library/hash_info.h` work? (Since it's specific to MD.)

Btw, it should probably be renamed `mbedtls_md_error_from_psa()` since it translates to MD error, and for consistency for our naming scheme (the `psa_` namespace is not for internal things) and the existing `mbedtls_pk_error_from_psa()`.

File: library/rsa.c (hunk 3 above)
Comment (mpg) [reply library/rsa.c:518]: They're equivalent in practice, but I agree checking for `NONE` looks a bit better.

File: library/rsa.c (hunk 2 above)
Comment (mpg) [reply library/rsa.c:63]: The novelty (use of PSA) is only in builds that where never possible before, so I think it's quite OK from a compatibility perspective.

I don't think we ever meant to promise that nothing would use PSA when `MBEDTLS_USE_PSA` is not defined, but I agree that this is a reasonable way to interpret the current working, so the new behaviour could be surprising to users. I think we can handle that by adding a warning every time something that previously depended on MD now starts depending on "MD or PSA", similar to [this one](https://github.com/Mbed-TLS/mbedtls/pull/6141/files#diff-0279e5915ac49daea5fd9ed14934a074d1a88e9b35c674d7d1a0d0e0688f60bbR1146) - so, eventually X.509 and TLS will get such a warning. And probably we should tune the wording on `MBEDTLS_USE_PSA` as well before the next release to accurately reflect the status as this point - I'll track that as part of #6146.

File: library/rsa.c (hunk 4 above)
Comment (AndrzejKurek) [reply library/rsa.c:1102]: I was thinking about using it globally across Mbed TLS, hence the `int module` argument - this way there would some module-specific errors first (in a `switch( module )`, followed by a fallback - presumably, set of shared error codes that are the same for every module.

File: include/mbedtls/mbedtls_config.h (hunk 1 above)
Comment (mpg) [reply include/mbedtls/mbedtls_config.h:1144]: Now, there's nothing specific about MD5 here. However, I agree that we should clarify that if using MD, you also need the `MBEDTLS_xxx_C` for all the hashes you can to use, and otherwise you also need `PSA_WANT_ALG_xxx` for those hashes.

File: library/rsa.c (hunk 4 above)
Comment (mpg) [reply library/rsa.c:1102]: Oh, sorry, I read too fast and missed that part. Yes, why not, it there's some code that can be shared between modules, we could save some code size that way. But I'm slightly worried that the function would become big and hard to read. Also, sometimes even within a module you need variants, see `mbedtls_pk_error_from_psa_rsa()` which may return either RSA or PK error codes.

I'm inclined to keep this focused on MD as a first step, then look into consolidation with other modules only later, unless you feel strongly otherwise.

File: tests/scripts/all.sh [hunk 5]
```diff
@@ -1760,6 +1759,8 @@ component_test_psa_crypto_config_accel_rsa_signature () {
 
     scripts/config.py -f tests/include/test/drivers/config_test_driver.h set MBEDTLS_SHA1_C
     scripts/config.py -f tests/include/test/drivers/config_test_driver.h set MBEDTLS_SHA512_C
+    # We need to define either MD_C or all of the PSA_WANT_ALG_SHAxxx.
+    scripts/config.py -f tests/include/test/drivers/config_test_driver.h set MBEDTLS_MD_C
```
Comment (AndrzejKurek) [inline tests/scripts/all.sh:1763]: Since it's the PSA crypto config with acceleration test - wouldn't we want to go the PSA way to be closer to MD being handled by drivers as well?

File: library/rsa.c [hunk 6]
```diff
@@ -1140,8 +1201,115 @@ static int mgf_mask( unsigned char *dst, size_t dlen, unsigned char *src,
...
+        goto exit;
+
+exit:
+    mbedtls_md_free( &md_ctx );
+
+    return( ret );
+#else
+    psa_hash_operation_t op = PSA_HASH_OPERATION_INIT;
+    psa_algorithm_t alg = mbedtls_psa_translate_md( md_alg );
+    psa_status_t status = PSA_SUCCESS;
```
Comment (AndrzejKurek) [inline library/rsa.c:1260]: I'm not sure if it's still canon, but initializing to an error was a good practice before :)

File: library/rsa.c [hunk 7]
```diff
@@ -1088,29 +1098,68 @@ int mbedtls_rsa_private( mbedtls_rsa_context *ctx,
...
+
+    mbedtls_md_init( &md_ctx );
+    if( ( ret = mbedtls_md_setup( &md_ctx, md_info, 0 ) ) != 0 )
+        goto exit;
+
+    hlen = mbedtls_md_get_size( md_info );
+#else
+    psa_hash_operation_t op = PSA_HASH_OPERATION_INIT;
+    psa_algorithm_t alg = mbedtls_psa_translate_md( md_alg );
+    psa_status_t status = PSA_SUCCESS;
```
Comment (AndrzejKurek) [inline library/rsa.c:1155]: Initializing to error remark same as in other places

File: library/rsa.c (hunk 4 above)
Comment (AndrzejKurek) [reply library/rsa.c:1102]: Absolutely no strong feelings, it's just a thing that came to my mind, as I can see that we're duplicating similar code in a couple of places.

**Comment by AndrzejKurek** [comment]
I finished the first round of my review and it seems that things look good :)

File: library/rsa.c (hunk 4 above)
Comment (mpg) [reply library/rsa.c:1102]: Agreed. Whoever adopts this PR, I suggest moving this function to `library/hash_info.[ch]` under the name `mbedtls_md_error_from_psa()`. That way, #6146, #6147 and #6149 can use it too. It doesn't solve the problem of cross-module near-duplication, but at least it will avoid exact-duplication for all cases where we silently fall back to PSA in the absence of MD.

Regarding the larger picture, @AndrzejKurek can you create an issue about that so we don't forget about it? I guess the "Use PSA: misc. gaps" EPIC would be a good place.

**Comment by mpg** [comment]
I'm unassigning myself and labeling the PR as "needs: adoption" as I'll be away from now until next end of August.

File: library/rsa.c (hunk 2 above)
Comment (mpg) [reply library/rsa.c:63]: I just thought: anyway, the only case where PSA is used in RSA is when MD is not defined, and the only way you'll be able to compile TLS or X.509 without MD is by enabling `USE_PSA` (they will depend on `MD_C || USE_PSA`). So, X.509 and TLS are indeed only going to use PSA when `USE_PSA` is enabled.

File: tests/scripts/all.sh (hunk 5 above)
Comment (mprse) [reply tests/scripts/all.sh:1763]: From the commit message I think this way was picked as it is simpler at least for now:

>Previously MD_C was auto-enabled based on the fact that ALG_RSA_PSS was
requested, but that's no longer the case since the previous commit.
We can fix this in one of two ways: either enable MD_C, or enable all
the PSA_WANT_ALG_SHA_xxx that are needed for test. Go for MD_C because
it's a single line and avoids having to enumerate a list that might grow
in the future.

Maybe we should have psa and no psa versions now?

File: library/rsa.c (hunk 6 above)
Comment (mprse) [reply library/rsa.c:1260]: Looking at the code I think that `status` can be simply initialized to `PSA_ERROR_CORRUPTION_DETECTED` without any other changes.

As far as I remember from review of my PRs it is acceptable to initialize `status` to `PSA_SUCCESS`, but code path must be clear and the comment should be added.

File: library/rsa.c (hunk 7 above)
Comment (mprse) [reply library/rsa.c:1155]: It seems that here it make sense to leave it as it is now since `ret` is also initialized to `0` and as you mentioned is is intended to handle `dlen==0` case.

File: library/rsa.c (hunk 3 above)
Comment (mprse) [reply library/rsa.c]: Fixed as suggested.

File: library/rsa.c (hunk 4 above)
Comment (mprse) [reply library/rsa.c]: Renamed function `ret_from_status->mbedtls_md_error_from_psa` and moved to library/hash_info.[ch] as global function.

File: tests/scripts/all.sh (hunk 5 above)
Comment (mprse) [reply tests/scripts/all.sh:1763]: Left current version.

File: library/rsa.c (hunk 6 above)
Comment (mprse) [reply library/rsa.c]: Fixed as suggested + initialized also `ret` to error value.

File: library/rsa.c (hunk 7 above)
Comment (mprse) [reply library/rsa.c:1136]: Left current version as it make sense in this case.

**Comment by mprse** [comment]
Addressed review comments from @AndrzejKurek .

File: library/rsa.c (hunk 4 above)
Comment (AndrzejKurek) [reply library/rsa.c]: Issue added here: https://github.com/Mbed-TLS/mbedtls/issues/6158 with a remark to start by assesing if the potential gain is worth it.

**Review by AndrzejKurek — APPROVED** [review]

**Comment by mprse** [comment]
This PR still needs additional reviewer as I adopted the PR, so I think in this case I shouldn't review it.
We need it for further work on https://github.com/orgs/Mbed-TLS/projects/1#column-18883163 tasks.

**Review by wernerlewis — COMMENTED** [review]
Changes LGTM, merge conflicts due to the removal of the radix arg in one of the test cases.

**Comment by mprse** [comment]
I couldn't rebase and force push probably this PR comes from @mpg fork repository.
I was able to resolve conflicts via github. Let's see if CI passes and review again.

File: tests/suites/test_suite_pk.data [hunk 8]
```diff
@@ -438,36 +438,36 @@ depends_on:MBEDTLS_RSA_C:MBEDTLS_PKCS1_V15:MBEDTLS_GENPRIME
...
+depends_on:MBEDTLS_PKCS1_V21:MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA
+pk_rsa_verify_ext_test_vec:"c0719e9a8d5d838d861dc6f675c899d2b309a3a65bb9fe6b11e5afcbf9a2c0b1":MBEDTLS_MD_SHA256:1024:16:"00dd118a9f99bab068ca2aea3b6a6d5997ed4ec954e40deecea07da01eaae80ec2bb1340db8a128e891324a5c5f5fad8f590d7c8cacbc5fe931dafda1223735279461abaa0572b761631b3a8afe7389b088b63993a0a25ee45d21858bab9931aedd4589a631b37fcf714089f856549f359326dd1e0e86dde52ed66b4a90bda4095":16:"010001":"0d2bdb0456a3d651d5bd48a4204493898f72cf1aaddd71387cc058bc3f4c235ea6be4010fd61b28e1fbb275462b53775c04be9022d38b6a2e0387dddba86a3f8554d2858044a59fddbd594753fc056fe33c8daddb85dc70d164690b1182209ff84824e0be10e35c379f2f378bf176a9f7cb94d95e44d90276a298c8810f741c9":MBEDTLS_PK_RSASSA_PSS:MBEDTLS_MD_SHA256:MBEDTLS_RSA_SALT_LEN_ANY:128:0
```
Comment (wernerlewis) [inline tests/suites/test_suite_pk.data:442]: The radix `16` args shouldn't be reintroduced in these test cases, as these args were removed from the `pk_rsa_verify_ext_test_vec` function. The extra args are then causing the CI to fail.

Comment (mprse) [reply tests/suites/test_suite_pk.data]: Yes, just pushed a fixed and tested locally.

**Review by wernerlewis — DISMISSED** [review]

File: tests/suites/test_suite_pk.data [hunk 9]
```diff
@@ -438,6 +438,7 @@ depends_on:MBEDTLS_RSA_C:MBEDTLS_PKCS1_V15:MBEDTLS_GENPRIME
 pk_rsa_alt:
 
 Verify ext RSA #1 (PKCS1 v2.1, salt_len = ANY, OK)
+
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_pk.data]: Unnecessary whitespace

**Comment by AndrzejKurek** [comment]
I'm not sure if we're fine with such merge commits; Wouldn't it be cleaner to create a new PR with cherry-picked commits and fixed history?
Apart from that - looks good to me.

**Comment by mprse** [comment]
We have here entire PR history, so this merge commit maybe is not so bad.
Let's wait for the CI.

**Comment by gilles-peskine-arm** [comment]
We generally try to avoid merge commits in pull requests, but they aren't forbidden. If the PR already has had significant review activity, and rebasing is not easy, a merge can be the better solution. At a glance, in this case, a merge seems like  a good solution.

I do have a request however: please document the merge conflict resolution in the commit message of the merge.

**Comment by mprse** [comment]
I rebased the PR on top of development as we normally do. PR was already reviewed and accepted by both reviewers, so this should not be a problem.
Previously I wasn't able to do a force push probably because of network connection problem.

**Review by wernerlewis — APPROVED** [review]

**Review by AndrzejKurek — APPROVED** [review]
LGTM
